### ISSUE TITLE
EMA with buffers to stabilize the teacher

### ISSUE THREAD

**Issue by petwu** [saltoricristiano/cosmix-uda#21]
Hi,

first of all, thanks for your great work.
I was experimenting with it for my thesis and notices that there is a small but significant bug in the EMA update of the teacher. You update only the parameters, but forget the model **buffers**. Without this, the teachers batch norm statistics aren't updated and they become increasingly misaligned with the other weights with every step leading to a degrading teacher performance over time.

A simple fix would look like this:
```python
class MomentumUpdater:
    ...

    @torch.no_grad()
    def update(self, online_net: nn.Module, momentum_net: nn.Module):
        """Performs the momentum update for each param group.
        Args:
            online_net (nn.Module): online network (e.g. online backbone, online projection, etc...).
            momentum_net (nn.Module): momentum network (e.g. momentum backbone,
                momentum projection, etc...).
        """

        for op, mp in zip(online_net.parameters(), momentum_net.parameters()):
            mp.data.mul_(self.cur_tau)
            mp.data.add_(op.data * (1 - self.cur_tau))

        # THIS IS THE IMPORTANT PART
        for ob, mb in zip(online_net.buffers(), momentum_net.buffers()):
            if ob.is_floating_point():
                ob.data.mul_(self.cur_tau)
                ob.data.add_(mb.data * (1 - self.cur_tau))
            else:
                ob.data.copy_(mb.data)
```

In my experiments, with this fix applied the teacher improves in a stable manner over time and eventually converges. It also makes the whole training less sensitive to the EMA update interval, allows for larger improvement through DA and also allows to use the teacher during inference/validation. I always thought it was a bit weird that your work and other papers like SALUDA reported the student's performance, rather than the teacher's as usual, until I find this bug, though it took me a while to find it ...

Hope this helps anyone that might stumble upon this 😊

**Comment by saltoricristiano** [comment]
Hi @petwu , thank you for posting this issue. 
I would like to follow more on this but I don't have the time. However, I would have a few additional questions:
- Would you mind sharing more on the obtained results with this fix? 
- Could you provide a comparison between teacher's results with/without the fix?
- And, does the fix enable even better results also for the student network or just for the teacher?
Thx!

**Comment by petwu** [comment]
Tbh, my thesis is already a few months back and I almost forgot opening this issue, therefore I don't have all numbers at hand anymore. But I'll try to answer your questions as best as I can.

Given this setup:
- Cylinder3D backbone with single linear layer as segmentation head (grid size of [500, 360, 80] for radius, azimuth angle and height, respectively)
- multi-target domain adaptation
  - source domain: SemanticKITTI (SK)
  - target domains: SemanticPOSS (SP), nuScenes (NS), PandaSet (PS) and Waymo (WO)
- data loading:
  - limit points to [-50,50] meters in x/y and [-5, 3] in z direction
  - labels mapped to a common subset of 6 classes (vehicle, bike, person, ground, manmade, vegetation)
  - augmentations: random rotation/scaling/translation/flipping, point jittering, PolarMix+LaserMix
- pretraining: 45k iterations, AdamW, lr=1e-4 (w/ consine annealing to 2e-5)
- adaptation: 32k iterations, AdamW, lr=2e-5 (fixed), EMA 0.999@1, no PolarMix/LaserMix augmentation

I got these mIoU results:
|                                         |   Model |    SK |    SP |    NS |    PS |    WO |  Avg. |       |
| :-------------------------------------- | ------: | ----: | ----: | ----: | ----: | ----: | ----: | ----: |
| (1) source-only                         |       - | 80.53 | 61.38 | 58.02 | 70.68 | 74.73 | 69.07 |       |
| (2a) EMA w/o buffers (best, 17k iters)  | student | 82.04 | 65.29 | 63.01 | 74.95 | 75.57 | 72.17 | +3.10 |
| (2b) EMA w/o buffers (final, 32k iters) | student | 80.47 | 59.90 | 53.25 | 64.34 | 71.27 | 65.84 | -3.23 |
| (3) EMA w/ buffers (final, 32k iters)   | teacher | 81.92 | 66.87 | 67.98 | 77.56 | 76.81 | 74.23 | +5.16 |

<img width="1011" height="511" alt="Image" src="https://github.com/user-attachments/assets/ae1b3b9c-1091-4f7e-86b7-b2ca8b02ae08" />

> Note: The right graphs has weird looking steps, because I kept the teacher fixed for the first 1k iterations to stabilize the student s. t. batch normalization statistics and parameters can adapt to the mixed domain before updating the teacher to prevent early performance degradation. After that, the teacher is updated once with a hard copy before starting with EMA updates.

Unfortunately, I only have the numbers for student  vs teacher at hand not teacher vs teacher, but if I remember correctly, then the teacher in (2) was mostly worse than the student, so the results still hold. Also, the student and teacher in (3) were roughly on par but with the student performance being more noisy/less stable.

As you can see, the teacher is much more stable than the student and including the buffers in the EMA update helps prevent the drop in performance over time. This allows for larger improvements and also makes the adaptation less sensitive to the number of iterations hyperparameter.
